src/frontend/hooks/useMapStyle.ts

```diff
@@ -0,0 +1,154 @@
+import * as React from "react";
+import MapboxGL from "@react-native-mapbox-gl/maps";
+import AsyncStorage from "@react-native-async-storage/async-storage";
+
+import api from "../api";
+import { useNetInfo } from "@react-native-community/netinfo";
+import ky from "ky";
+import { normalizeStyleURL } from "../lib/mapbox";
+import config from "../../config.json";
+import { useExperiments } from "./useExperiments";
+
+/** Key used to store most recent map id in Async Storage */
+const MAP_STYLE_KEY = "@MAPSTYLE";
+
+/** URL used for map style when no custom map and user is online */
+const onlineStyleURL = MapboxGL.StyleURL.Outdoors + "?" + Date.now();
+
+/** URL used for map style when user is not online
+ * generated by [mapeo-offline-map](https://github.com/digidem/mapeo-offline-map) */
+const fallbackStyleURL = "asset://offline-style.json";
+
+type LegacyMapActions =
+  | { action: "setLoading" }
+  | { action: "setCustom"; styleUrl: string }
+  | { action: "setOnline" }
+  | { action: "setFallback" };
+
+type LegacyMapTypes =
+  | { type: "custom" | "online" | "fallback"; styleUrl: string }
+  | { type: "loading"; styleUrl: null };
+
+export type MapTypes =
+  | "loading"
+  | "mapServer"
+  | "custom"
+  | "online"
+  | "fallback";
+
+function legacyMapTypeReducer(
+  state: LegacyMapTypes,
+  action: LegacyMapActions
+): LegacyMapTypes {
+  switch (action.action) {
+    case "setCustom":
+      return { type: "custom", styleUrl: action.styleUrl };
+    case "setOnline":
+      return { type: "online", styleUrl: onlineStyleURL };
+    case "setFallback":
+      return { type: "fallback", styleUrl: fallbackStyleURL };
+    default:
+      return { type: "loading", styleUrl: null };
+  }
+}
```

Comment:
Right, Is my assumption here that: if custom maps are available, we should serve that, and the online state doesnt matter? That is why im using a reducer here, because knowing the online state is not always necessary

Reply:
^or is that assumption wrong?

Reply:
I think that assumption is correct. But it means when maintaining this code you need to know that you should only call "setOnline()" when mapType !== 'custom', which can lead to bugs. I feel it kind of obfuscates what is happening, I think what you actual have here is a function like the one in my comment above. To expand that code above:

```ts
const {type, styleUrl} = getStyleUrl(onlineState, customMapState)

function getStyleUrl(
  onlineState: 'unknown' | 'online' | 'offline',
  customMapState: 'unknown' | 'available' | 'unavailable'
) {
  if (onlineState === 'unknown' || customMapState === 'unknown')  {
    return { type: 'loading', styleUrl: null }
  } else if (customMapState === 'available') {
    return { type: 'custom', styleUrl: customMapStyleUrl }
  } else if (onlineState === 'online') {
    return { type: "online", styleUrl: onlineStyleURL };
  } else {
      return { type: "fallback", styleUrl: fallbackStyleURL };
  }
}
```

I think writing it that way you avoid double-renders (render, then call useEffect, then render), you avoid the risk of render loops, you avoid needing to know which actions you can dispatch based on the current state, and in my opinion it is clearer for the reader what the result is given each state (e.g. I can easily see from the above that if the custom map is available then that will be what is used).

Reply:
(this is similar to what is in the original hook, just cleaned up a bit)